### PR TITLE
nonRoot feature, with server split into server and engine.

### PR DESCRIPTION
**Pull request by palladia**
@Microsoft/omi-devs

### CODE REVIEW

File: Unix/protocol/protocol.c [hunk 1]
```diff
@@ -440,6 +469,96 @@ static StrandFT _ProtocolSocket_FT = {
...
+
+static MI_Result _ProtocolSocketTrackerRemoveElement(Sock s)
+{
+    TrackerBucket b;
+    int r;
+
+    b.key = s;
+
+    Lock_Acquire(&s_trackerLock);
+    r = HashMap_Remove(&s_protocolSocketTracker, (HashBucket*)&b);
```
Comment (paulcallen) [inline Unix/protocol/protocol.c]: how does the bucket get deleted?

File: Unix/protocol/protocol.c [hunk 2]
```diff
@@ -440,6 +469,96 @@ static StrandFT _ProtocolSocket_FT = {
...
+    b->key = s;
+    b->value = p;
+
+    Lock_Acquire(&s_trackerLock);
+    r = HashMap_Insert(&s_protocolSocketTracker, (HashBucket*)b);
+    Lock_Release(&s_trackerLock);
+
+    if (r != 0)
+    {
+        // already exists
```
Comment (paulcallen) [inline Unix/protocol/protocol.c]: can insert fail?
Also cant hash maps cause hash collisions. I assume exist does actually mean it exists, not just the hash that exists.

File: Unix/protocol/protocol.c [hunk 3]
```diff
@@ -440,6 +469,96 @@ static StrandFT _ProtocolSocket_FT = {
...
+        trace_TrackerHashMapAlreadyExists(p, s);
+        MI_Result res = _ProtocolSocketTrackerRemoveElement(s);
+        if (res == MI_RESULT_OK)
+        {
+            Lock_Acquire(&s_trackerLock);
+            r = HashMap_Insert(&s_protocolSocketTracker, (HashBucket*)b);
```
Comment (paulcallen) [inline Unix/protocol/protocol.c]: can it fail?

File: Unix/protocol/protocol.c [hunk 4]
```diff
@@ -490,8 +615,16 @@ static MI_Boolean _SendAuthRequest(
...
-    req->uid = geteuid();
-    req->gid = getegid();
+    if (uid != INVALID_ID && gid != INVALID_ID)
+    {
+        req->uid = uid;
+        req->gid = gid;
+    }
+    else
+    {
+        req->uid = geteuid();
```
Comment (paulcallen) [inline Unix/protocol/protocol.c]: which process are we in here? Why would we get invalid ids here? it does not seem to make sense to do this

File: Unix/protocol/protocol.c (hunk 1 above)
Comment (palladia) [inline Unix/protocol/protocol.c]: HashMap_Remove() does it.

File: Unix/protocol/protocol.c (hunk 2 above)
Comment (palladia) [inline Unix/protocol/protocol.c]: The only failure case is if r == 1, which means that the bucket exists.

File: Unix/protocol/protocol.c (hunk 3 above)
Comment (palladia) [inline Unix/protocol/protocol.c]: It could.  And in that case, this AddElement function will return MI_RESULT_FAILED.

File: Unix/protocol/protocol.c (hunk 4 above)
Comment (palladia) [inline Unix/protocol/protocol.c]: Only client would call _SendAuthRequest with invalid ids.

File: Unix/build.mak [hunk 5]
```diff
@@ -102,7 +102,7 @@ endif
 ifdef CONFIG_SINGLEIMAGE
   DIRECTORIES += image
 else
-  DIRECTORIES += server agent
+  DIRECTORIES += server agent engine
```
Comment (jeffaco) [inline Unix/build.mak]: So it looks like there's a new binary here, but it hasn't been put into the kit. This means that the RPM and DPKG kits that we build won't be valid, and will install a non-working version of OMI.

Note that PBUILD currently only builds --dev, not kits. Building kits are done on a restricted set of machines (just SLES 10 for Linux) due to universal binaries. You minimally need to make the `installbuilder` changes, build new kits on a SLES 10 system, and then test it to make sure that the version you install actually works.

I'm going to need this to be done prior to signing off this change.

File: Unix/engine/GNUmakefile [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+##
+## Note: on HPUX_PARISC_HP, the server must be linked as C++, since we have
+## not found a way to dynamically load a C++ provider from a C program.
+##
+
+ifdef HPUX_PARISC_HP
+  CXXPROGRAM = omiengine
+  SOURCES = engine.c enginemaincxx.cpp ../server/servercommon.c
+else
+  CPROGRAM = omiengine
```
Comment (jeffaco) [inline Unix/engine/GNUmakefile]: So binary file `omiengine` should now exist in `/opt/omi/bin`?

File: Unix/engine/engine.c [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    /* Extract configuration file options */
+    GetConfigFileOptions();
+
+    /* Extract command-line options a second time (to override) */
+    GetCommandLineOptions(&argc, argv);
+
+    /* Open the log file */
+    OpenLogFile();
+
+#if defined(CONFIG_POSIX)
```
Comment (jeffaco) [inline Unix/engine/engine.c]: Hmm. When would we NOT be using POSIX? What's the purpose if this `#if`?

File: Unix/engine/engine.c [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        }
+
+        r = SendSocketFileRequest(&s_data.protocol1->protocolSocket);
+        if (r == MI_FALSE)
+        {
+            err(ZT("failed to send socket file request"));
+        }
+
+        // Give it a little time for SocketFile info to come back from server
+        // Sleep_Milliseconds(50);
```
Comment (jeffaco) [inline Unix/engine/engine.c]: This is here why? Do we need extra time or not? What if the system is heavily loaded?

File: Unix/engine/engine.c [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+**
+**==============================================================================
+*/
+
+#include <server/server.h>
+
+static Options s_opts;
+static ServerData s_data;
+
+int enginemain(int argc, const char* argv[])
```
Comment (jeffaco) [inline Unix/engine/engine.c]: No common header file, and this wasn't extern "C"'d.  What happens if `Unix/engine/enginemaincxx.cpp` is used (I guess on HP-UX, for whatever reason)?

I'm guessing you didn't test this on that platform? Does it even pass `PBUILD`?

File: Unix/etc/omiserver.conf [hunk 10]
```diff
@@ -58,3 +58,5 @@ NoSSLCompression=false
 #NtlmCredsFile=PATH
 # For example
 #NtlmCredsFile=/etc/opt/omi/.creds/ntlm
+nonroot=false
+service=test1
```
Comment (jeffaco) [inline Unix/etc/omiserver.conf]: I guess this is a "fake" omiserver.conf file (not the one we ship to customers). But still, do we really want lines 61/62 here? Shouldn't unit tests set this up if that's what they want?

File: Unix/protocol/protocol.c [hunk 11]
```diff
@@ -8,6 +8,9 @@
 */
 
 #include <assert.h>
+#include <sys/types.h>
```
Comment (jeffaco) [inline Unix/protocol/protocol.c]: I don't fully understand this code ... @yakman2020 can you please review this too? I'd like to see Bruce review this prior to checkin please.

File: Unix/provreg/provreg.c [hunk 12]
```diff
@@ -511,7 +511,7 @@ static int _AddEntry(
 #endif /* defined(CONFIG_ENABLE_PREEXEC) */
 
     /* ProvRegEntry.hosting*/
-    e->hosting = PROV_HOSTING_INPROC;
+    e->hosting = PROV_HOSTING_REQUESTOR;
```
Comment (jeffaco) [inline Unix/provreg/provreg.c]: So is this changing the default hosting type to `@requestor@`, assuming that the `HOSTING=` line isn't specified?

Where is that parsed from the .reg file? Is `inproc` disabled now? Or will it be a separate commit?

If a separate commit, what's the purpose of the changes here?

File: Unix/provreg/provreg.c [hunk 13]
```diff
@@ -528,7 +528,7 @@ static int _AddEntry(
          */
         if (strcmp(hosting, PROV_REG_HOSTING_INPROC) == 0)
         {
-            e->hosting = PROV_HOSTING_INPROC;
+            e->hosting = PROV_HOSTING_REQUESTOR;
```
Comment (jeffaco) [inline Unix/provreg/provreg.c]: Yuck. Why not just make INPROC be invalid, and return an error? I see no reason why asking, explicitly, for INPROC should not be an error and would do something completely different ...

Please explain your rationale, thanks.

File: Unix/tests/wsman/utils.cpp [hunk 14]
```diff
@@ -308,6 +309,9 @@ int StartServerAndConnect(
 
     UT_ASSERT( MI_RESULT_OK == result );
 
+    // A new socket connection is created, give it a little time
+    ut::sleep_ms(200);
```
Comment (jeffaco) [inline Unix/tests/wsman/utils.cpp]: I hate this kind of thing. What if the test system is slower than you expect, for whatever reason? Why not wait for the socket for some "reasonable" period of time, like this:

```
int tries = 0;
while ( true )
{
    if (socket_exists)
    {
        break;
    }

    if (++tries > 10);
    {
        emit error and abort;
    }

    ut:sleep_ms(100);
}
```

This is much more resilient, even if the machine is slower than expected for whatever reason.

Comment (jeffaco) [inline Unix/tests/wsman/utils.cpp]: This sort of thing is true for other hard-coded sleep statements above this too ...

File: Unix/etc/omiserver.conf (hunk 10 above)
Comment (jeffaco) [inline Unix/etc/omiserver.conf]: Should defaults be added to the "real" omiserver.conf file shipping to customers on a kit install?

File: Unix/build.mak (hunk 5 above)
Comment (palladia) [inline Unix/build.mak]: Fixed installbuilder/datafiles/Base_OMI.data

File: Unix/engine/GNUmakefile (hunk 6 above)
Comment (palladia) [inline Unix/engine/GNUmakefile]: yes

File: Unix/engine/engine.c (hunk 9 above)
Comment (palladia) [inline Unix/engine/engine.c]: pbuild tested on HP

File: Unix/engine/engine.c (hunk 7 above)
Comment (palladia) [inline Unix/engine/engine.c]: leftover from days of supporting Windows.  If we want to remove Windows support, we can go in and clean them up.

File: Unix/provreg/provreg.c [hunk 15]
```diff
@@ -615,7 +616,7 @@ static int _AddEntryForExtraClass(
          */
         if (strcmp(regClass->hosting, PROV_REG_HOSTING_INPROC) == 0)
         {
-            hosting = PROV_HOSTING_INPROC;
+            hosting = PROV_HOSTING_REQUESTOR;
```
Comment (paulcallen) [inline Unix/provreg/provreg.c]: Overriding inproc silently does not sound right. You definitely need a very big error telling which provider asked for inproc, but it was made out of processes and that the provider probably needs to be updated.

File: Unix/http/httpauth.c [hunk 16]
```diff
@@ -1639,19 +1711,52 @@ MI_Boolean IsClientAuthorized(_In_ Http_SR_SocketData * handler)
...
+                trace_HTTP_UserAuthFailed("user not authenticated");
+                handler->authFailed = TRUE;
+                
+                _SendAuthResponse(handler, auth_response, response_len);
+                return FALSE;
+            }
+        }
+        else
+        {
+            if (0 != _AskServerToAuthenticateUser(headers->username,
```
Comment (paulcallen) [inline Unix/http/httpauth.c]: OK, I think we need to make this asynchronous and use the binary protocol properly, sending a proper PamAuthResponse from the server and break this code apart so the code after this point happens as part of the PanAuthResponse.
We need to do this because we are on the main Protocol_Run thread and we are blocking that thread from processing other requests during this time

File: Unix/http/httpauth.c [hunk 17]
```diff
@@ -1639,19 +1711,52 @@ MI_Boolean IsClientAuthorized(_In_ Http_SR_SocketData * handler)
...
+            if (0 != AuthenticateUser(headers->username, headers->password))
+            {
+                handler->httpErrorCode = HTTP_ERROR_CODE_UNAUTHORIZED;
+                auth_response = (char *)RESPONSE_HEADER_UNAUTH_FMT;
+                response_len  = RESPONSE_HEADER_UNAUTH_FMT_LEN;
+
+                trace_HTTP_UserAuthFailed("user not authenticated");
+                handler->authFailed = TRUE;
+                
+                _SendAuthResponse(handler, auth_response, response_len);
```
Comment (palladia) [inline Unix/http/httpauth.c]: Not sure which failure path you are referring to.  These blocks of code deal with failure cases.  If you mean _SendAuthResponse...it's a void function, but does log if it fails.
